virtio-devices/src/balloon.rs

```diff
@@ -201,6 +203,30 @@ impl BalloonEpollHandler {
 
                 let gpa = (pfn as u64) << VIRTIO_BALLOON_PFN_SHIFT;
                 if let Ok(hva) = mem.get_host_address(GuestAddress(gpa)) {
+                    if queue_index == 0 {
+                        let region = mem.find_region(GuestAddress(gpa)).unwrap();
+                        let (host_fd, f_offset) = if let Some(f_off) = region.file_offset() {
+                            (f_off.file().as_raw_fd(), f_off.start())
+                        } else {
+                            (-1, 0)
+                        };
+
+                        if host_fd != -1 {
+                            let offset = hva as usize - region.as_ptr() as usize;
+                            let res = unsafe {
+                                libc::fallocate64(
+                                    host_fd,
+                                    libc::FALLOC_FL_PUNCH_HOLE | libc::FALLOC_FL_KEEP_SIZE,
+                                    (offset as u64 + f_offset) as libc::off64_t,
+                                    (1 << VIRTIO_BALLOON_PFN_SHIFT) as libc::off64_t,
+                                )
+                            };
+                            if res != 0 {
+                                return Err(Error::FallocateFail(io::Error::last_os_error()));
+                            }
+                        }
```

Comment:
This whole block can be rewritten:
```rust
                        if let Some(f_off) = region.file_offset() {
                            let offset = hva as usize - region.as_ptr() as usize;
                            let res = unsafe {
                                libc::fallocate64(
                                    f_off.file().as_raw_fd(),
                                    libc::FALLOC_FL_PUNCH_HOLE | libc::FALLOC_FL_KEEP_SIZE,
                                    (offset as u64 + f_off.start()) as libc::off64_t,
                                    (1 << VIRTIO_BALLOON_PFN_SHIFT) as libc::off64_t,
                                )
                            };
                            if res != 0 {
                                return Err(Error::FallocateFail(io::Error::last_os_error()));
                            }
                        }
```

Reply:
Thank you. I will rewrite this.

Reply:
Cool, let me when the PR is ready for another review!